packages/mattermost-redux/src/selectors/entities/channels.ts

```diff
@@ -879,6 +922,108 @@ export const getUnreadChannels: (state: GlobalState, lastUnreadChannel?: Channel
+    getAllChannels,
+    getAllTeamsUnreadChannelIds,
+    getTeammateNameDisplaySetting,
+    (currentUser, profiles, userIdsInChannels, channels, allTeamsUnreadChannelIds, settings) => {
+        // If we receive an unread for a channel and then a mention the channel
+        // won't be sorted correctly until we receive a message in another channel
+        if (!currentUser) {
+            return [];
+        }
+
+        return allTeamsUnreadChannelIds.filter((id) => channels[id] && channels[id].delete_at === 0).map((id) => {
+            const c = channels[id];
+
+            if (c.type === General.DM_CHANNEL || c.type === General.GM_CHANNEL) {
+                return completeDirectChannelDisplayName(currentUser.id, profiles, userIdsInChannels[id], settings!, c);
+            }
+
+            return c;
+        });
+    },
+);
+
+export const sortUnreadChannels = (
+    channels: Channel[],
+    myMembers: RelationOneToOne<Channel, ChannelMembership>,
+    lastPosts: RelationOneToOne<Channel, Post>,
+    lastUnreadChannel: LastUnreadChannel | null,
+    crtEnabled: boolean,
+) => {
+    function isMuted(channel: Channel) {
+        return isChannelMuted(myMembers[channel.id]);
+    }
+
+    function hasMentions(channel: Channel) {
+        if (lastUnreadChannel && channel.id === lastUnreadChannel.id && lastUnreadChannel.hadMentions) {
+            return true;
+        }
+
+        const member = myMembers[channel.id];
+        return member?.mention_count !== 0;
+    }
+
+    // Sort channels with mentions first and then sort by recency
+    return [...channels].sort((a, b) => {
+        // Sort muted channels last
+        if (isMuted(a) && !isMuted(b)) {
+            return 1;
+        } else if (!isMuted(a) && isMuted(b)) {
+            return -1;
+        }
+
+        // Sort non-muted mentions first
+        if (hasMentions(a) && !hasMentions(b)) {
+            return -1;
+        } else if (!hasMentions(a) && hasMentions(b)) {
+            return 1;
+        }
+
+        let lastPostAt = {
+            a: a.last_post_at,
+            b: b.last_post_at,
+        };
+
+        if (crtEnabled) {
+            lastPostAt = {
+                a: a.last_root_post_at,
+                b: b.last_root_post_at,
+            };
+        }
+
+        // If available, get the last post time from the loaded posts for the channel, but fall back to the
+        // channel's last_post_at if that's not available. The last post time from the loaded posts is more
+        // accurate because channel.last_post_at is not updated on the client as new messages come in.
+        const aLastPostAt = maxDefined(lastPostAt.a, lastPosts[a.id]?.create_at);
+        const bLastPostAt = maxDefined(lastPostAt.b, lastPosts[b.id]?.create_at);
+
+        return bLastPostAt - aLastPostAt;
+    });
+};
+
+export const getSortedAllTeamsUnreadChannels: (state: GlobalState, lastUnreadChannel?: LastUnreadChannel | null) => Channel[] = createSelector(
```

Comment:
To come back around to this, can you remove the `lastUnreadChannel` argument from this? I don't think we should need it any more, but it being there means we could still break the memoization on this